### PR TITLE
Adding null-conditional operator on IEnumerable.GetEnumerator()

### PR DESCRIPTION
**Pull request by smaglio81**
Summary of the changes (Less than 80 chars)
 - Added `?.` operator to prevent null reference exception (`_items?.GetEnumerator()`)

Addresses #16938

### CODE REVIEW

**Comment by BrennanConroy** [comment]
Looks like there can also be a null ref in `CopyTo`
https://github.com/aspnet/AspNetCore/blob/d66db3b0fffdcb010914a4d5a8cb4426f5f1c7ce/src/Http/Http/src/Internal/ItemsDictionary.cs#L98-L104

File: src/Http/Http/test/Internal/ItemsDictionaryTests.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    public class ItemsDictionaryTests
+    {
+        [Fact]
+        public void GetEnumerator_ShouldResolveWithoutNullReferenceException()
+        {
+            // Arrange
+            var dict = new Microsoft.AspNetCore.Http.ItemsDictionary();
+
+            // Act and Assert
+            System.Collections.IEnumerable en = (System.Collections.IEnumerable) dict;
```
Comment (BrennanConroy) [inline src/Http/Http/test/Internal/ItemsDictionaryTests.cs]: Add `using System.Collections;`

File: src/Http/Http/test/Internal/ItemsDictionaryTests.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace Microsoft.AspNetCore.Http
+{
+    public class ItemsDictionaryTests
+    {
+        [Fact]
+        public void GetEnumerator_ShouldResolveWithoutNullReferenceException()
+        {
+            // Arrange
+            var dict = new Microsoft.AspNetCore.Http.ItemsDictionary();
```
Comment (BrennanConroy) [inline src/Http/Http/test/Internal/ItemsDictionaryTests.cs]: Add `using Microsoft.AspNetCore.Http;`

File: src/Http/Http/test/Internal/ItemsDictionaryTests.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+            // Act and Assert
+            System.Collections.IEnumerable en = (System.Collections.IEnumerable) dict;
+            foreach(var item in en) // <-- in the original code, the implicit call to .GetEnumerator() would throw a NullReferenceException
+            {
+                // if there is a problem this code won't be reached
+            }
+
+            Assert.True(true, "The code should make it here without throwing an exception.");
```
Comment (BrennanConroy) [inline src/Http/Http/test/Internal/ItemsDictionaryTests.cs]: No need for this

File: src/Http/Http/test/Internal/ItemsDictionaryTests.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        [Fact]
+        public void GetEnumerator_ShouldResolveWithoutNullReferenceException()
+        {
+            // Arrange
+            var dict = new Microsoft.AspNetCore.Http.ItemsDictionary();
+
+            // Act and Assert
+            System.Collections.IEnumerable en = (System.Collections.IEnumerable) dict;
+            foreach(var item in en) // <-- in the original code, the implicit call to .GetEnumerator() would throw a NullReferenceException
```
Comment (BrennanConroy) [inline src/Http/Http/test/Internal/ItemsDictionaryTests.cs]: Just simplify to `en.GetEnumerator();`?

File: src/Http/Http/test/Internal/ItemsDictionaryTests.cs (hunk 2 above)
Comment (Tratcher) [reply src/Http/Http/test/Internal/ItemsDictionaryTests.cs]: This test class is already in that namespace, you shouldn't need the using or the prefix.

File: src/Http/Http/test/Internal/ItemsDictionaryTests.cs (hunk 4 above)
Comment (Tratcher) [reply src/Http/Http/test/Internal/ItemsDictionaryTests.cs]: `Assert.NotNull(en.GetEnumerator());`

File: src/Http/Http/test/Internal/ItemsDictionaryTests.cs (hunk 1 above)
Comment (smaglio81) [reply src/Http/Http/test/Internal/ItemsDictionaryTests.cs]: Can do!

File: src/Http/Http/test/Internal/ItemsDictionaryTests.cs (hunk 2 above)
Comment (smaglio81) [reply src/Http/Http/test/Internal/ItemsDictionaryTests.cs]: Thanks!

File: src/Http/Http/test/Internal/ItemsDictionaryTests.cs (hunk 3 above)
Comment (smaglio81) [reply src/Http/Http/test/Internal/ItemsDictionaryTests.cs]: Removed, no problem.

File: src/Http/Http/test/Internal/ItemsDictionaryTests.cs (hunk 4 above)
Comment (smaglio81) [reply src/Http/Http/test/Internal/ItemsDictionaryTests.cs]: Nice!

**Comment by smaglio81** [comment]
Okay, I've attempted to integrate the recommended updates. Let me know how it looks.

**Comment by BrennanConroy** [comment]
Are you interested in fixing `CopyTo` also?

**Comment by smaglio81** [comment]
In the code snippet, it seemed like line 98 was doing a null check? Also, I'm kind of new to the github resolution system ... since that comment didn't have the conversational display (and resolution button) like the other comments I thought it might have been a momentary concern that was self-resolved.

Let me know if you still want it updated after reviewing line 98?

https://github.com/aspnet/AspNetCore/blob/d66db3b0fffdcb010914a4d5a8cb4426f5f1c7ce/src/Http/Http/src/Internal/ItemsDictionary.cs#L98-L104

**Comment by BrennanConroy** [comment]
> line 98 was doing a null check?

Yeah, but the after the null check line 104 uses `_items` which would null ref.

**Comment by smaglio81** [comment]
Great Catch! Since it seems like .CopyTo wants to copy the items into the dictionary, what if around line 102 EnsureDictionary() was called to initialize `_items`?

**Comment by BrennanConroy** [comment]
It's copying from `_items` to the passed in array, we don't want to allocate the dictionary.

I think just adding an `else` would be good.

**Comment by smaglio81** [comment]
Can do!

**Comment by smaglio81** [comment]
So, I started going down the route of using an `else` statement, but I wanted to format the curly braces in the way the code normally does it. After poking around for an example, it started to seem like `else` statements weren't used a lot. So, I went back to your original suggestion and used the null-conditional operator.

I also added a unit test for it. I'm not sure if I hit the standards on it.

**Review by BrennanConroy — APPROVED** [review]

**Comment by BrennanConroy** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=422315&view=logs

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/2730
https://github.com/aspnet/AspNetCore-Internal/issues/3241

**Comment by BrennanConroy** [comment]
Thanks @smaglio81 !

**Pull request merged**